### PR TITLE
fix: [sessiond] only unsuspend credits that have non 0 quota

### LINKED ISSUE #6993: Gy Redirect Policy Rule removed after Validity Timer triggered CCR-U with FUA Redirect
### Your Environment

- **Version:** 1.4 orc8r + FeG, AGW Bare Metal 1.4.0-1618930512-8dcd7a4e
- **Affected Component:** FeG + Access Gateway
- **Affected Subcomponent:** Sessiond / Session Proxy
- **Deployment Environment:** e.g. AWS (orc8r), bare-metal (AGW)

### Describe the Issue

We are regression testing 1.4 as we upgrade from 1.3. We are testing UE attach on SIM with zero balance in OCS. We expect that CCR-I is replied with FUI/FUA redirect. We have redirectd enabled on the gateway. 

On attach, UE is connected, has zero balance, and redirect works. When validity time expires, the CCA-U has FUA redirect and policy rule for redirect is deactivated at which point UE can consume traffic and it isn't accounted as it still has no quota. 

**To Reproduce**
1. Setup feg_lte network with FeG Gy + Gx relay
2. Provision a SIM in OCS with zero balance
3. Attach UE to network
4. Observe CCA-I contains FUI/FUA redirect
5. try to surf to http://wumo.com/wumo and see that redirect works
6. leave UE idle until validity timer expires and triggers CCR-U 
7. observe CCA-U contains FUI/FUA redirect
8. Observe in sessiond logs that redirect policy rule gets deactivated and leaves open rule "rg1_high_speed" in place
9. try to surf any website and see that data access is possible. 

**Expected behavior**
A CCA-U with FUA redirect should add or keep the redirect policy rule in place, it should not remove the redirect policy rule regardless of GSU value. 

**Key Log Snippet**
```
May 18 06:33:02 agw sessiond[2807]: I0518 06:33:02.710675  2807 SessionState.cpp:497] Updating used charging credit for Rule=rg1_high_speed Rating Group=1 Service Identifier=0
May 18 06:33:12 agw sessiond[2807]: I0518 06:33:12.709408  2824 GrpcMagmaUtils.cpp:38]
May 18 06:33:12 agw sessiond[2807]:   magma.lte.RuleRecordTable {
May 18 06:33:12 agw sessiond[2807]:       records {
May 18 06:33:12 agw sessiond[2807]:         sid: "IMSI315010999900108"
May 18 06:33:12 agw sessiond[2807]:         rule_id: "internal_default_drop_flow_rule"
May 18 06:33:12 agw sessiond[2807]:         ue_ipv4: "192.168.128.14"
May 18 06:33:12 agw sessiond[2807]:       }
May 18 06:33:12 agw sessiond[2807]:       records {
May 18 06:33:12 agw sessiond[2807]:         sid: "IMSI315010999900108"
May 18 06:33:12 agw sessiond[2807]:         rule_id: "rg1_high_speed"
May 18 06:33:12 agw sessiond[2807]:         ue_ipv4: "192.168.128.14"
May 18 06:33:12 agw sessiond[2807]:       }
May 18 06:33:12 agw sessiond[2807]:       records {
May 18 06:33:12 agw sessiond[2807]:         sid: "IMSI315010999900108"
May 18 06:33:12 agw sessiond[2807]:         rule_id: "redirect"
May 18 06:33:12 agw sessiond[2807]:         ue_ipv4: "192.168.128.14"
May 18 06:33:12 agw sessiond[2807]:       }
May 18 06:33:12 agw sessiond[2807]:       epoch: 1621318255
May 18 06:33:12 agw sessiond[2807]:   }
May 18 06:33:12 agw sessiond[2807]: I0518 06:33:12.710125  2807 SessionState.cpp:497] Updating used charging credit for Rule=rg1_high_speed Rating Group=1 Service Identifier=0
May 18 06:33:12 agw sessiond[2807]: I0518 06:33:12.710211  2807 LocalSessionManagerHandler.cpp:87] Sending 1 charging updates and 0 monitor updates to OCS and PCRF
May 18 06:33:12 agw sessiond[2807]: I0518 06:33:12.712232  2807 GrpcMagmaUtils.cpp:38]
May 18 06:33:12 agw sessiond[2807]:   magma.lte.UpdateSessionRequest {
May 18 06:33:12 agw sessiond[2807]:       updates {
May 18 06:33:12 agw sessiond[2807]:         usage {
May 18 06:33:12 agw sessiond[2807]:           bytes_tx: 632
May 18 06:33:12 agw sessiond[2807]:           bytes_rx: 3234
May 18 06:33:12 agw sessiond[2807]:           charging_key: 1
May 18 06:33:12 agw sessiond[2807]:           type: VALIDITY_TIMER_EXPIRED
May 18 06:33:12 agw sessiond[2807]:           requested_units {
May 18 06:33:12 agw sessiond[2807]:           }
May 18 06:33:12 agw sessiond[2807]:         }
May 18 06:33:12 agw sessiond[2807]:         session_id: "IMSI315010999900108-399779"
May 18 06:33:12 agw sessiond[2807]:         request_number: 1
May 18 06:33:12 agw sessiond[2807]:         sid: "IMSI315010999900108"
May 18 06:33:12 agw sessiond[2807]:         ue_ipv4: "192.168.128.14"
May 18 06:33:12 agw sessiond[2807]:         spgw_ipv4: "10.0.2.1"
May 18 06:33:12 agw sessiond[2807]:         apn: "internet"
May 18 06:33:12 agw sessiond[2807]:         plmn_id: "315010"
May 18 06:33:12 agw sessiond[2807]:         imsi_plmn_id: "315010"
May 18 06:33:12 agw sessiond[2807]:         user_location: "\202\023\005\020\000\001\023\005\020\000\006\"\001"
May 18 06:33:12 agw sessiond[2807]:         tgpp_ctx {
May 18 06:33:12 agw sessiond[2807]:           gx_dest_host: "gx00.i-new.com"
May 18 06:33:12 agw sessiond[2807]:           gy_dest_host: "gy00.inew-cs.com"
May 18 06:33:12 agw sessiond[2807]:         }
May 18 06:33:12 agw sessiond[2807]:         common_context {
May 18 06:33:12 agw sessiond[2807]:           sid {
May 18 06:33:12 agw sessiond[2807]:             id: "IMSI315010999900108"
May 18 06:33:12 agw sessiond[2807]:           }
May 18 06:33:12 agw sessiond[2807]:           ue_ipv4: "192.168.128.14"
May 18 06:33:12 agw sessiond[2807]:           apn: "internet"
May 18 06:33:12 agw sessiond[2807]:           teids {
May 18 06:33:12 agw sessiond[2807]:             enb_teid: 1532801
May 18 06:33:12 agw sessiond[2807]:             agw_teid: 4
May 18 06:33:12 agw sessiond[2807]:           }
May 18 06:33:12 agw sessiond[2807]:         }
May 18 06:33:12 agw sessiond[2807]:       }
May 18 06:33:12 agw sessiond[2807]:   }
May 18 06:33:14 agw sessiond[2807]: I0518 06:33:14.154769  2807 GrpcMagmaUtils.cpp:38]
May 18 06:33:14 agw sessiond[2807]:   magma.lte.UpdateSessionResponse {
May 18 06:33:14 agw sessiond[2807]:       responses {
May 18 06:33:14 agw sessiond[2807]:         success: true
May 18 06:33:14 agw sessiond[2807]:         sid: "IMSI315010999900108"
May 18 06:33:14 agw sessiond[2807]:         charging_key: 1
May 18 06:33:14 agw sessiond[2807]:         credit {
May 18 06:33:14 agw sessiond[2807]:           validity_time: 900
May 18 06:33:14 agw sessiond[2807]:           is_final: true
May 18 06:33:14 agw sessiond[2807]:           final_action: REDIRECT
May 18 06:33:14 agw sessiond[2807]:           granted_units {
May 18 06:33:14 agw sessiond[2807]:             total {
May 18 06:33:14 agw sessiond[2807]:             }
May 18 06:33:14 agw sessiond[2807]:             tx {
May 18 06:33:14 agw sessiond[2807]:             }
May 18 06:33:14 agw sessiond[2807]:             rx {
May 18 06:33:14 agw sessiond[2807]:             }
May 18 06:33:14 agw sessiond[2807]:           }
May 18 06:33:14 agw sessiond[2807]:           redirect_server {
May 18 06:33:14 agw sessiond[2807]:             redirect_address_type: URL
May 18 06:33:14 agw sessiond[2807]:             redirect_server_address: "http://www.sersimple.com/PortalCautivo.php"
May 18 06:33:14 agw sessiond[2807]:           }
May 18 06:33:14 agw sessiond[2807]:         }
May 18 06:33:14 agw sessiond[2807]:         result_code: 2001
May 18 06:33:14 agw sessiond[2807]:         tgpp_ctx {
May 18 06:33:14 agw sessiond[2807]:           gx_dest_host: "gx00.i-new.com"
May 18 06:33:14 agw sessiond[2807]:           gy_dest_host: "gy00.inew-cs.com"
May 18 06:33:14 agw sessiond[2807]:         }
May 18 06:33:14 agw sessiond[2807]:         session_id: "IMSI315010999900108-399779"
May 18 06:33:14 agw sessiond[2807]:       }
May 18 06:33:14 agw sessiond[2807]:   }
May 18 06:33:14 agw sessiond[2807]: I0518 06:33:14.155704  2807 ChargingGrant.cpp:87] GSU for RG: 1 IMSI315010999900108-399779 is invalid, but accepting it as it has a final unit action or suspended credit
May 18 06:33:14 agw sessiond[2807]: I0518 06:33:14.155721  2807 SessionCredit.cpp:574] Received a GSU with no valid grants, keeping the same type
May 18 06:33:14 agw sessiond[2807]: I0518 06:33:14.155727  2807 SessionCredit.cpp:189] Received the following credit total_volume=0 tx_volume=0 rx_volume=0 grant_tracking_type=TRACKING_UNSET
May 18 06:33:14 agw sessiond[2807]: I0518 06:33:14.155738  2807 ChargingGrant.cpp:346] IMSI315010999900108-399779 received a credit RG: 1 with final action REDIRECT, redirect_server: http://www.sersimple.com/PortalCautivo.php with expiry timer in 900 seconds
May 18 06:33:14 agw sessiond[2807]: I0518 06:33:14.155781  2807 GrpcMagmaUtils.cpp:38]
May 18 06:33:14 agw sessiond[2807]:   magma.lte.DeactivateFlowsRequest {
May 18 06:33:14 agw sessiond[2807]:       sid {
May 18 06:33:14 agw sessiond[2807]:         id: "IMSI315010999900108"
May 18 06:33:14 agw sessiond[2807]:       }
May 18 06:33:14 agw sessiond[2807]:       rule_ids: "redirect"
May 18 06:33:14 agw sessiond[2807]:       request_origin {
May 18 06:33:14 agw sessiond[2807]:         type: GY
May 18 06:33:14 agw sessiond[2807]:       }
May 18 06:33:14 agw sessiond[2807]:       ip_addr: "192.168.128.14"
May 18 06:33:14 agw sessiond[2807]:   }

```
**Additional context**
Gy pcap, session_proxy log, sessiond full log
[Cyan200.zip](https://github.com/magma/magma/files/6504782/Cyan200.zip)

### PR DESCRIPTION
**Pull request by themarwhal**
## Summary
See attached issue for the reported problem.

In short, we have to make sure that a credit has >0 quota before un-suspending. 
We currently only check for these two things:
1. credit received is valid
2. credit is in suspension

But since we could receive a valid credit with 0 grant, we should also check that the there is quota available.

The logic is slightly messy as we handle 'suspended' credits slightly differently from credits that have been deactivated due to FUA-redirect/restrict. 

In order to make this work, I've had to make two changes:

When determining whether to remove Gy rules, check that the credit needs to be activated. This is done by the logic check:
```
// in update_charging_credits
bool should_activate = session->is_credit_ready_to_be_activated(credit_key);
if (!should_activate) {
  continue;
}
...
if (prev_final_action) {
  // uninstall gy rules
}
```
When receiving a new credit, determine if the credit needs to be deactivated. We didn't check for this before, but this is useful for the edge case where we receive a FUA-terminate for an already empty quota.

## Unit tests added
`test_no_credit_reauth_with_redirected_suspended_credit` (fails on master w/o the changes)
`test_terminate_credit_reauth_with_redirected_suspended_credit` (fails on master w/o the changes)

## Test Plan
make precommit_sm

the new unit test fails on master w/o this change

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Review by uri200 — COMMENTED** [review]
note that there is a difference of being suspended and being with credit 0

If we receive credit, but the credit is 0 and we generate traffic, I guess our session should be terminated.

However, with those changes we will keep the session. 

So imagine a situation were the subscriber does not  have data anymore, PCRF keeps sending grants 0 (not sure if the PCRF would be so dumb), but the subscriber is never allowed to leave.

Not sure if this situation is even possible, but maybe worth thinking on it before landing this

**Comment by themarwhal** [comment]
> note that there is a difference of being suspended and being with credit 0
> 
> If we receive credit, but the credit is 0 and we generate traffic, I guess our session should be terminated.
> 
> However, with those changes we will keep the session.
> 
> So imagine a situation were the subscriber does not have data anymore, PCRF keeps sending grants 0 (not sure if the PCRF would be so dumb), but the subscriber is never allowed to leave.
> 
> Not sure if this situation is even possible, but maybe worth thinking on it before landing this

@uri200 

Yeah well this case is it is already suspended AND credit is 0. I think in this case we don't want to unsuspend the credit even if the OCS keeps sending 0.

Although the case we might not handling might be:
1. we receive a credit with FUA-redirect/restrict
2. we use up our credit and go into redirect/restrict aka suspension
3. we get an update by expiry timer / RAR still with 0 grant but now with fua-terminate

I think in this case hopefully we terminate the session. But who knows :D 

I'll write a unit test and see what happens in that case.

File: lte/gateway/c/session_manager/LocalEnforcer.cpp [hunk 1]
```diff
@@ -1187,69 +1187,83 @@ void LocalEnforcer::add_rules_for_unsuspended_credit(
...
+                 << result_code << " during update" << session_id;
+    actions->suspended_credits.insert(
+        ImsiSessionIDAndCreditkey{imsi, session_id, ckey});
+  } else {
+    // TODO( ): deal with other error codes
+    MLOG(MERROR) << "Received Unimplemented result code " << result_code
+                 << " for " << session_id << " during update. Not action taken";
+  }
+  return true;
+}
```
Comment (uri200) [inline lte/gateway/c/session_manager/LocalEnforcer.cpp:1219]: I remember I tried with a similar function and I felt that the function was mixing too many things, doing stuff in some objects, and returning a value. That is why I prefered to had the logic on `update_charging_credits`

File: lte/gateway/c/session_manager/SessionState.cpp [hunk 2]
```diff
@@ -1630,6 +1630,9 @@ bool SessionState::receive_charging_credit(
                 << " Activating service RG: " << key << " for " << session_id_;
     grant->set_service_state(SERVICE_NEEDS_ACTIVATION, credit_uc);
   }
+  if (grant->should_deactivate_service()) {
+    grant->set_service_state(SERVICE_NEEDS_DEACTIVATION, credit_uc);
+  }
```
Comment (uri200) [inline lte/gateway/c/session_manager/SessionState.cpp:1635]: why do we check it here. Don't we have to wait for the next cycle to check?

Comment (themarwhal) [reply lte/gateway/c/session_manager/SessionState.cpp:1635]: @uri200 This case will only cover the case where the credit is already exhausted and FUA-terminate is set. Since at this point no new usage is added, it just acts on the terminate instruction. 
(I don't think it makes much sense to wait until the next cycle in this case)

File: lte/gateway/c/session_manager/LocalEnforcer.cpp (hunk 1 above)
Comment (themarwhal) [reply lte/gateway/c/session_manager/LocalEnforcer.cpp:1219]: Yeah it's a bit opaque but tried to make it so that the boolean value returned just indicates whether we want to further process the credit or not. 
The `update_charging_credits` function was getting too long/ hard to read :/

Comment (themarwhal) [reply lte/gateway/c/session_manager/LocalEnforcer.cpp:1219]: I can add a comment on the line where we call the function too, if that helps with clarity.

**Comment by uri200** [comment]
I would recommend to run this on TeraVM with tests
```
gxgy05.ntl
gxgy05_capacity.ntl
gxgy06.ntl
gxgy06_capacity.ntl
```
(5 is gy credit, and 6 is gx monitors)

example
```
~/magma/automation$ ng40test gxgy05_capacity.ntl
```

**Review by uri200 — APPROVED** [review]

File: lte/gateway/c/session_manager/LocalEnforcer.cpp (hunk 1 above)
Comment (uri200) [reply lte/gateway/c/session_manager/LocalEnforcer.cpp:1219]: yes, maybe a comment will help us in the future.

**Comment by themarwhal** [comment]
> I would recommend to run this on TeraVM with tests
> 
> ```
> gxgy05.ntl
> gxgy05_capacity.ntl
> gxgy06.ntl
> gxgy06_capacity.ntl
> ```
> 
> (5 is gy credit, and 6 is gx monitors)
> 
> example
> 
> ```
> ~/magma/automation$ ng40test gxgy05_capacity.ntl
> ```
Ran the tests
Screen Shot 2021-05-24 at 1 59 32 PM

File: lte/gateway/c/session_manager/LocalEnforcer.cpp (hunk 1 above)
Comment (themarwhal) [reply lte/gateway/c/session_manager/LocalEnforcer.cpp:1219]: will do!

**Comment by talkhasib** [comment]
is this getting backported to 1.5? 1.4?

**Comment by themarwhal** [comment]
> is this getting backported to 1.5? 1.4?

@talkhasib yes, but it takes a bit of work. Will put up a PR against 1.5 tomorrow. (Will have to see how difficult backporting to 1.4 will be)

**Comment by talkhasib** [comment]
> > is this getting backported to 1.5? 1.4?
> 
> @talkhasib yes, but it takes a bit of work. Will put up a PR against 1.5 tomorrow. (Will have to see how difficult backporting to 1.4 will be)

Lets only backport to 1.5 for now
